src/widget/wtrackmenu.cpp

```diff
@@ -0,0 +1,1548 @@
+    addToAutoDJ(PlaylistDAO::AutoDJSendLoc::BOTTOM);
+}
+
+void WTrackMenu::slotAddToAutoDJTop() {
+    addToAutoDJ(PlaylistDAO::AutoDJSendLoc::TOP);
+}
+
+void WTrackMenu::slotAddToAutoDJReplace() {
+    addToAutoDJ(PlaylistDAO::AutoDJSendLoc::REPLACE);
+}
+
+void WTrackMenu::addToAutoDJ(PlaylistDAO::AutoDJSendLoc loc) {
+    if (m_pTrackIdList.empty()) {
+        return;
+    }
+    const TrackIdList trackIds = m_pTrackIdList;
+    if (trackIds.isEmpty()) {
+        qWarning() << "No tracks selected for AutoDJ";
+        return;
+    }
+
+    PlaylistDAO& playlistDao = m_pTrackCollectionManager->internalCollection()->getPlaylistDAO();
+
+    // TODO(XXX): Care whether the append succeeded.
+    m_pTrackCollectionManager->unhideTracks(trackIds);
+    playlistDao.addTracksToAutoDJQueue(trackIds, loc);
+}
+
+
+void WTrackMenu::slotCoverInfoSelected(const CoverInfoRelative& coverInfo) {
+    if (m_pTrackPointerList.empty()) {
+        return;
+    }
+    for (const auto pTrack : m_pTrackPointerList) {
+        pTrack->setCoverInfo(coverInfo);
+    }
+}
+
+void WTrackMenu::slotReloadCoverArt() {
+    if (m_pTrackPointerList.empty()) {
+        return;
+    }
+    guessTrackCoverInfoConcurrently(m_pTrackPointerList);
+}
+
+void WTrackMenu::slotPurge() {
+    DEBUG_ASSERT(m_pTrackModel);
+    if (m_pTrackModel) {
+        QModelIndexList indices = m_pSelectedTrackIndices;
+        if (indices.size() > 0) {
+            TrackModel* trackModel = m_pTrackModel;
+            if (trackModel) {
+                trackModel->purgeTracks(indices);
+            }
+        }
+    }
+}
+
+void WTrackMenu::slotUnhide() {
+    if (m_pTrackModel) {
+        QModelIndexList indices = m_pSelectedTrackIndices;
+
+        if (indices.size() > 0) {
+            TrackModel* trackModel = m_pTrackModel;
+            if (trackModel) {
+                trackModel->unhideTracks(indices);
+            }
+        }
+    }
+}
+
+void WTrackMenu::clearTrackSelection() {
+    m_pTrackIdList.clear();
+    m_pTrackPointerList.clear();
+    m_pSelectedTrackIndices.clear();
+}
+
+bool WTrackMenu::optionIsEnabled(Filter flag) {
+    bool optionIsAvailable = m_eFilters.testFlag(Filter::None) || m_eFilters.testFlag(flag);
+    bool optionIsValid = false;
+
+    if (!optionIsAvailable) {
+        return false;
+    }
+
+    Filters independentOptions =
+            Filter::Playlist |
+            Filter::Crate |
+            Filter::FileBrowser;
+
+    // Some of these can be made independent of track table.
+    Filters trackTableDependentOptions =
+            Filter::AutoDJ |            
+            Filter::LoadTo |
+            Filter::Remove |
+            Filter::Metadata |
+            Filter::Reset |
+            Filter::BPM |
+            Filter::Color |
+            Filter::HideUnhidePurge |
+            Filter::Properties;
+
+    if (independentOptions.testFlag(flag)) {
+        return true;
+    }
+
+    if (trackTableDependentOptions.testFlag(flag)) {
```

Comment:
move these checks to the constructor when refactoring to pass the TrackModel* to the constructor